initrd/etc/functions

```diff
@@ -759,61 +804,97 @@ prompt_new_owner_password() {
 	tpm_owner_password=1
 	tpm_owner_password2=2
 	while [ "$tpm_owner_password" != "$tpm_owner_password2" ] || [ "${#tpm_owner_password}" -gt 32 ] || [ -z "$tpm_owner_password" ]; do
-		read -s -p "New TPM Owner Password (2 words suggested, 1-32 characters max): " tpm_owner_password
-		echo
-
-		read -s -p "Repeat chosen TPM Owner Password: " tpm_owner_password2
-		echo
+		read -r -s -p $'\nNew TPM Owner Password (2 words suggested, 1-32 characters max): ' tpm_owner_password
+		read -r -s -p $'\nRepeat chosen TPM Owner Password: ' tpm_owner_password2
 
 		if [ "$tpm_owner_password" != "$tpm_owner_password2" ]; then
-			echo "Passphrases entered do not match. Try again!"
 			echo
+			echo "Passphrases entered do not match. Try again!"
 		fi
+		echo
 	done
 
 	# Cache the password externally to be reused by who needs it
 	DEBUG "Caching TPM Owner Password to /tmp/secret/tpm_owner_password"
 	mkdir -p /tmp/secret || die "Unable to create /tmp/secret"
-	echo -n "$tpm_owner_password" >/tmp/secret/tpm_owner_password || die "Unable to cache TPM password under /tmp/secret"
+	echo -n "$tpm_owner_password" >/tmp/secret/tpm_owner_password || die "Unable to cache TPM password under /tmp/secret/tpm_owner_password"
 }
 
 check_tpm_counter() {
+	# $1: rollback file path
 	TRACE_FUNC
 
 	LABEL=${2:-3135106223}
 	tpm_password="$3"
 	# if the /boot.hashes file already exists, read the TPM counter ID
 	# from it.
 	if [ -r "$1" ]; then
-		TPM_COUNTER=$(grep counter- "$1" | cut -d- -f2)
+		# Robustly extract the first hex string after 'counter-' on any line
+		TPM_COUNTER=$(grep -Eo 'counter-[0-9a-fA-F]+' "$1" | sed -n 's/counter-//p' | head -n1 | tr -d '\n')
```

Comment:
This is where check_tpm_counter failed and where TPM passphtase was asked in silenced output @marmarek